benchmark/atoi.cpp

```diff
@@ -52,16 +58,96 @@ auto lemire_as_zoo_swar(const char *chars) {
+    },
+    ExpressedAsEscapeCodes = { ' ', '\r', '\f', '\v', '\n', '\t' };
+    static_assert(SpaceCharacters == ExpressedAsEscapeCodes); */
+    using S = swar::SWAR<8, uint64_t>;
+    constexpr S Space{meta::BitmaskMaker<uint64_t, ' ', 8>::value};
+    auto space = swar::equals(bytes, Space);
+    auto otherWhiteSpace =
+        swar::constantIsGreaterEqual<'\r'>(bytes) &
+        ~swar::constantIsGreaterEqual<'\t' - 1>(bytes);
+    auto whiteSpace = space | otherWhiteSpace;
+    auto notWhiteSpace = S{S::MostSignificantBit} ^ whiteSpace;
+    return notWhiteSpace.lsbIndex();
+}
+
+/// @brief Loads the "block" containing the pointer, by proper alignment
+/// @tparam PtrT Pointer type for loading
+/// @tparam Block as the name indicates
+/// @param pointerInsideBlock the potentially misaligned pointer
+/// @param b where the loaded bytes will be put
+/// @return a pair to indicate the aligned pointer to the base of the block
+/// and the misalignment, in bytes, of the source pointer
+template<typename PtrT, typename Block>
+std::tuple<PtrT *, int>
+blockAlignedLoad(PtrT *pointerInsideBlock, Block *b) {
+    uintptr_t asUint = reinterpret_cast<uintptr_t>(pointerInsideBlock);
+    constexpr auto Alignment = alignof(Block), Size = sizeof(Block);
+    static_assert(Alignment == Size);
+    auto misalignment = asUint % Alignment;
+    auto *base = reinterpret_cast<PtrT *>(asUint - misalignment);
+    memcpy(b, base, Size);
+    return { base, misalignment };
+}
+
+/// \brief Helper function to fix the non-string part of block
+template<typename S>
+S adjustMisalignmentFor_strlen(S data, int misalignment) {
+    // The speculative load has the valid data in the higher lanes.
+    // To use the same algorithm as the rest of the implementation, simply
+    // populate with ones the lower part, in that way there won't be nulls.
+    constexpr typename S::type Zero{0};
+    auto
+        zeroesInMisalignedOnesInValid =
+            (~Zero) // all ones
+            <<  (misalignment * 8), // assumes 8 bits per char
+        onesInMisalignedZeroesInValid = ~zeroesInMisalignedOnesInValid;
+    return data | S{onesInMisalignedZeroesInValid};
+}
+
 std::size_t c_strLength(const char *s) {
-    using S = swar::SWAR<8, std::size_t>;
+    using S = swar::SWAR<8, uint64_t>;
     constexpr auto
         MSBs = S{S::MostSignificantBit},
         Ones = S{S::LeastSignificantBit};
-    S bytes;
-    for(auto base = s;; base += 8) {
-        memcpy(&bytes.m_v, base, 8);
+    constexpr auto BytesPerIteration = sizeof(S::type);
+    S initialBytes;
+
+    auto indexOfFirstTrue = [](auto bs) { return bs.lsbIndex(); };
+
+     // Misalignment must be taken into account because a SWAR read is
+    // speculative, it might read bytes outside of the actual string.
+    // It is safe to read within the page where the string occurs, and to
+    // guarantee that, simply make aligned reads because the size of the SWAR
+    // base size will always divide the memory page size
```

Comment:
Yes.  I have not yet created such a document.  Because this is introductory code we can keep it like this, right? once we have a body of improvements we can populate other parts of the library such as design documentation.
The most critical piece that only lives as a comment is the one describing associative iteration.

Reply:
Would it be valuable to set up a Doxygen generation from this and host it somewhere? 
Starting to have some documentation infra could motivate more and better documentation.

Reply:
Once in a blue moon I try Doxygen.  Sadly, it just does not understand my code.  I still use doxygen comments in the code.